pkg/kf/deploy.go

```diff
@@ -0,0 +1,82 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kf
+
+import (
+	"fmt"
+
+	"github.com/GoogleCloudPlatform/kf/pkg/kf/internal/envutil"
+	serving "github.com/knative/serving/pkg/apis/serving/v1alpha1"
+	cserving "github.com/knative/serving/pkg/client/clientset/versioned/typed/serving/v1alpha1"
+)
+
+// AppLister lists the deployed apps.
+type AppLister interface {
+	// List lists the deployed apps.
+	List(opts ...ListOption) ([]serving.Service, error)
+
+	// ListConfigurations the deployed configurations in a namespace.
+	ListConfigurations(...ListConfigurationsOption) ([]serving.Configuration, error)
+}
+
+// Deployer deploys an image to Knative. It will either create or update an
+// existing service. It should be created via NewDeployer().
+type Deployer interface {
+	Deploy(s serving.Service, opts ...DeployOption) (serving.Service, error)
+}
+
+type deployer struct {
+	c         cserving.ServingV1alpha1Interface
+	appLister AppLister
+}
+
+// NewDeployer creates a new Deployer.
+func NewDeployer(l AppLister, c cserving.ServingV1alpha1Interface) Deployer {
+	return &deployer{
+		appLister: l,
+		c:         c,
+	}
+}
+
+// Deploy deploys an image to Knative.
+func (d *deployer) Deploy(s serving.Service, opts ...DeployOption) (serving.Service, error) {
+	services, err := d.appLister.List(WithListAppName(s.Name))
```

Comment:
I think you can use the `ExtractOneService` wrapper here since you're looking for exactly one if I understand right